File: 2019/src/main/kotlin/day03/day3.kt
package day03

import java.io.File
import kotlin.math.abs

data class Quadruple<A, B, C, D>(val first: A, val second: B, val third: C, val fourth: D)

fun trace(path: String): Map<Pair<Int, Int>, Int> {
    val res = mutableMapOf<Pair<Int, Int>, Int>()
    var (y, x, distance) = Triple(0, 0, 0)

    for (step in path.split(",")) {
        val (dy, dx) = when (step[0]) {
            'U' -> -1 to 0
            'D' -> 1 to 0
            'R' -> 0 to -1
            'L' -> 0 to 1
            else -> throw IllegalArgumentException()
        }

        repeat(step.substring(1).toInt()) {
            y += dy
            x += dx
            distance++
            res.putIfAbsent(y to x, distance)
        }
    }

    return res
}

fun solve(paths: List<String>, distance: (Quadruple<Int, Int, Int, Int>) -> Int): Int {
    val trace1 = trace(paths[0])
    val trace2 = trace(paths[1])

    return trace1.keys.filter { it in trace2 }
        .minOf { distance(Quadruple(it.first, it.second, trace1[it]!!, trace2[it]!!)) }
}

fun evaluatorOne(input: List<String>): Int = solve(input) { abs(it.first) + abs(it.second) }
fun evaluatorTwo(input: List<String>): Int = solve(input) { it.third + it.fourth }

fun readLinesFromFile(filePath: String): Result<List<String>> =
    runCatching { File(filePath).readLines() }

fun main() {
    readLinesFromFile("src/main/resources/day03.txt")
        .onSuccess {
            println("Part One: ${evaluatorOne(it)}")
            println("Part Two: ${evaluatorTwo(it)}")
        }
        .onFailure {
            println("Error reading file: ${it.message}")
        }
}

File: 2019/src/main/scala/day03/day3.scala
package day03

import scala.util.{Try, Success, Failure, Using}
import scala.io.Source

case class Direction(dy: Int, dx: Int)

case class Point(y: Int, x: Int) {
    def abs = x.abs + y.abs
    def +(dir: Direction) = Point(y + dir.dy, x + dir.dx)
}

case class Pair(pos: Point, distance: Int)

type WirePath = Map[Point, Int]

def parseLine(line: String) = line.split(",").collect {
    case s"R$num" => (Direction(0, 1)  -> num.toInt)
    case s"D$num" => (Direction(1, 0)  -> num.toInt)
    case s"U$num" => (Direction(-1, 0) -> num.toInt)
    case s"L$num" => (Direction(0, -1) -> num.toInt)
}.toList

def parseInput(input: List[String]) = input.map(parseLine)

def preComputeTrace(path: List[(Direction, Int)]): WirePath = {
    val directions = path.flatMap { case (dir, amount) => List.fill(amount)(dir) }
    return directions.scanLeft(Point(0, 0))(_ + _).zipWithIndex.tail.distinctBy(_._1).toMap
}

def solve(traces: List[WirePath], distance: Pair => Int): Int = {
    val List(trace1, trace2) = traces
    return (trace1.keySet & trace2.keySet).map(it => distance(Pair(it, trace1(it) + trace2(it)))).min
}

def evaluatorOne(input: List[WirePath]) = solve(input, it => it.pos.abs)
def evaluatorTwo(input: List[WirePath]) = solve(input, it => it.distance)

def readLinesFromFile(filePath: String): Try[List[String]] =
    Using(Source.fromResource(filePath))(_.getLines().toList)

def hello(): Unit = {
    readLinesFromFile("day03.txt") match {
        case Success(lines) => {
            val traces = parseInput(lines).map(preComputeTrace)
            println(s"Part One: ${evaluatorOne(traces)}")
            println(s"Part Two: ${evaluatorTwo(traces)}")
        }
        case Failure(exception) => {
            println(s"Error reading file: ${exception.getMessage}")
        }
    }
}